keps/sig-api-machinery/storage-migration.md

```diff
@@ -0,0 +1,360 @@
+      * [Graduation Criteria](#graduation-criteria)
+      * [Alternatives](#alternatives)
+         * [“Active management”](#active-management)
+         * [update-storage-objects.sh](#update-storage-objectssh)
+
+
+## Summary
+
+We propose to extend and improve the [oc adm migrate storage][] command line
+tool to migrate the stored API objects in Kubernetes clusters. We will
+deliver a standalone tool of alpha quality in 2018 Q3. We will integrate the
+storage migration into the Kubernetes upgrade process in 2018 Q4. We will make
+the migration automatically triggered in 2019.
+
+[oc adm migrate storage]:https://www.mankier.com/1/oc-adm-migrate-storage
+
+## Motivation
+
+"Today it is possible to create API objects (e.g., HPAs) in one version of
+Kubernetes, go through multiple upgrade cycles without touching those objects,
+and eventually arrive at a version of Kubernetes that can’t interpret the stored
+resource and crashes. See k8s.io/pr/52185."[1][]. We propose a solution to the
+problem.
+
+[1]:https://docs.google.com/document/d/1eoS1K40HLMl4zUyw5pnC05dEF3mzFLp5TPEEt4PFvsM
+
+### Goals
+
+A successful storage version migration tool must:
+* work for Kubernetes built-in APIs, custom resources (CR), and aggregated APIs.
+* do not add burden to cluster administrators or Kubernetes distributions.
+* only cause insignificant load to apiservers. For example, if the master has
+  10GB memory, the migration tool should generate less than 10 qps of single
+  object operations(TODO: measure the memory consumption of PUT operations;
+  study how well the default 10 Mbps bandwidth limit in the oc command work).
+* work for big clusters that have ~10^6 instances of some resource types.
+* make progress in flaky environment, e.g., flaky apiservers, or the migration
+  process get preempted.
+* allow system administrators to track the migration progress.
+
+As to the deliverables,
+* in the short term, providing system administrators with a tool to migrate
+  the API objects to the proper storage versions.
+* in the long term, automating the migration of Kubernetes built-in APIs, CR,
+  aggregated APIs without further burdening system administrators or Kubernetes
+  distributions.
+
+## Proposal
+
+### Alpha workflow
+
+We will create a **kube-migrator** command line tool. At the alpha stage,
+**kube-migrator** needs to be manually invoked, and does not handle custom
+resources. 
+
+After all the apiservers are at the desired version, including kube-apiservers
+and aggregated apiservers, the cluster administrator launches **kube-migrator**.
+It
+* deletes any existing `deployment` of **kube-migrator controller** 
+* creates a `deployment` of **kube-migrator controller** 
+* uses the same kubeconfig file as **kubectl** does.
+
+On start-up, the **kube-migrator controller** does the following,
+* discovers all supported APIs.
+* lists all custom resource definitions, excludes them from the API list
```

Comment:
so only CRDs are excluded? what about aggregated apis?

Reply:
Aggregated apis will be included in the first iteration, see line 90.